src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,276 @@
+  private static final double   JITTER_MAX    = 0.20;
+  private static final String[] COUNTER_ARGS  = {
+          "OVERFLOW", "WRAP",
+          "INCRBY", "u32", "0", "1",
+          };
+
+  private final String              serverApiUrl;
+  private final ReplicatedJedisPool jedisPool;
+  private final AccountsManager     accountsManager;
+  private final Client              client;
+
+  private DirectoryReconciliationWorker directoryReconciliationWorker;
+
+  public DirectoryReconciler(ContactDiscoveryConfiguration cdsConfig,
+                             ReplicatedJedisPool jedisPool,
+                             AccountsManager accountsManager)
+  {
+    this.serverApiUrl = cdsConfig.getServerApiUrl();
+    this.jedisPool = jedisPool;
+    this.accountsManager = accountsManager;
+
+    SslConfigurator sslConfig = SslConfigurator.newInstance()
+                                               .securityProtocol("TLSv1.2");
+    try {
+      sslConfig.trustStore(initializeKeyStore(cdsConfig.getServerApiCaCertificate()));
+    } catch (CertificateException ex) {
+      logger.error("error reading serverApiCaCertificate from contactDiscovery config", ex);
+    }
+
+    this.client = ClientBuilder.newBuilder()
+                               .register(HttpAuthenticationFeature.basic("signal", cdsConfig.getServerApiToken().getBytes()))
+                               .sslContext(sslConfig.createSSLContext())
+                               .build();
+  }
+
+  private static KeyStore initializeKeyStore(String pemCaCertificate)
+          throws CertificateException
+  {
+    try {
+      PEMReader       reader      = new PEMReader(new InputStreamReader(new ByteArrayInputStream(pemCaCertificate.getBytes())));
+      X509Certificate certificate = (X509Certificate) reader.readObject();
+
+      if (certificate == null) {
+        throw new CertificateException("No certificate found in parsing!");
+      }
+
+      KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      keyStore.load(null);
+      keyStore.setCertificateEntry("ca", certificate);
+      return keyStore;
+    } catch (IOException | KeyStoreException ex) {
+      throw new CertificateException(ex.toString());
+    } catch (NoSuchAlgorithmException ex) {
+      throw new AssertionError(ex);
+    }
+  }
+
+  @Override
+  public void start() {
+    this.directoryReconciliationWorker = new DirectoryReconciliationWorker(serverApiUrl, jedisPool, accountsManager, client);
+
+    this.directoryReconciliationWorker.start();
+  }
+
+  @Override
+  public void stop() {
+    directoryReconciliationWorker.shutdown();
+  }
+
+  private static class DirectoryReconciliationWorker extends Thread {
+
+    private final String              serverApiUrl;
+    private final ReplicatedJedisPool jedisPool;
+    private final AccountsManager     accountsManager;
+    private final Client              client;
+
+    private final String workerId;
+    private final Random random;
```

Comment:
There's never a reason not to use SecureRandom

Reply:
done in 4d5b40bb442c40bcda99aaa1748b6581875479b8